### PR TITLE
Update blinky to use MCUboot with changes from Mbed OS team

### PR DESCRIPTION
**Pull request by LDong-Arm**
(Copied from https://github.com/LDong-Arm/mbed-mcuboot-blinky/pull/1)
## Intro
This updates Embedded Planet's blinky app for MCUboot.
Tested on DISCO_L475VG_IOT01A and NRF52840_DK.

**Note**: This is not a complete update to the blinky demo:
* README has not been updated
* Serial OTA has not been integrated - for now we preload the secondary image to a known place in the internal flash, and the primary app copies the image onto the secondary_bd (QSPI flash).

### Usage
To build:
* Erase the internal flash, to remove any residual data:
    ```
    pyocd erase --chip
    ```
* Build and flash the bootloader: https://github.com/AGlass0fMilk/mbed-mcuboot-demo/pull/5
* Build this blinky app
* Sign with the **rsa-2048** key you generated for the bootloader (unless you changed the configuration):
    ```
    mcuboot/scripts/imgtool.py sign -k <path-to-key>.pem --align 4 -v 1.0.0 --header-size 4096 --pad-header -S 0xC0000 ./BUILD/<TARGET>/GCC_ARM/mbed-mcuboot-blinky.bin v1.[bin|hex]
    ```
* Flash the primary image:
  * NRF52840_DK: drag & drop v1.hex onto the board
  * DISCO_L475VG_IOT01A:
    ```
    pyocd flash -a 0x8020000 v1.bin
    ```
* Increment `version` in `main.cpp` and rebuild.
* Sign and flash the secondary image:
  * NRF52840_DK: **add `--hex-addr 0x60000`** when signing v2.hex, the drag & drop v2.hex
  * DISCO_L475VG_IOT01A: flash the update image with an offset
    ```
    pyocd flash -a 0x8060000 v2.bin
    ```
    **Note**: The application will later copied the second image to QSPIFBlockDevice to emulate an OTA. This is _not_ a proper thing to do - maybe Embedded Planet can reintegrate their Serial OTA into the demo, to fetch a new image via serial instead.

To run:
* _First_ button press prompt:
    ```
    > Press button to erase secondary slot
    ```
    **Note**: This erases the QSPIFBlockDevice in case there's residual data.
* _Second_ button press prompt:
    ```
    > Press button to copy update image to secondary BlockDevice
    ```
    **Note**: This copies the image to QSPIFBlockDevice. As said above, it's a hack to speed up development.
* _Third_ button press prompt:
    ```
    > Image copied to secondary BlockDevice, press button to activate
    ```
    Press the button again, then reboot.
* This time MCUboot should perform the update.

### CODE REVIEW

File: mbed_app.json [hunk 1]
```diff
@@ -1,15 +1,60 @@
...
+            "demo-button-active-low": true,
+            "target.mbed_app_start": "0x8021000",
+            "target.mbed_app_size": "0xBE000",
+            "mcuboot.primary-slot-address": "0x8020000",
+            "mcuboot.slot-size": "0xC0000",
+            "mcuboot.scratch-address": "0x80E0000",
+            "mcuboot.scratch-size": "0x20000",
+            "mcuboot.max-img-sectors": "0x180",
+            "mcuboot.read-granularity": 4,
+            "qspif.QSPI_MIN_PROG_SIZE": 4
```
Comment (LDong-Arm) [inline mbed_app.json:57]:
```suggestion
            "mcuboot.read-granularity": 1,
            "qspif.QSPI_MIN_PROG_SIZE": 1
```
The Disco board allows one-byte access. See https://github.com/ARMmbed/mbed-os/pull/13848/files

File: mbed_app.json [hunk 2]
```diff
@@ -1,15 +1,60 @@
...
+        },
+        "DISCO_L475VG_IOT01A": {
+            "demo-button": "USER_BUTTON",
+            "demo-button-active-low": true,
+            "target.mbed_app_start": "0x8021000",
+            "target.mbed_app_size": "0xBE000",
```
Comment (AGlass0fMilk) [inline mbed_app.json:50]: Why is this `0xBE000` and not `0xBF000`? As it stands now, there is a 4kB gap between the end of the main application slot and the start of the scratch space region:

`target.mbed_app_start + target.mbed_app_size = 0x21000 + 0xBE000 = 0xDF000`
`0xE0000 - 0xDF000 = 0x1000`

Is this to allow space for something like KVStore? If not, should we change the default to `0xBF000` in this example to maximize application space?

**Review by AGlass0fMilk — COMMENTED** [review]
Just one question regarding the configuration.

File: mbed_app.json (hunk 2 above)
Comment (AGlass0fMilk) [reply mbed_app.json:50]: Here is a WIP diagram of the memory map I made for the readme of this demo:

mbed-mcuboot-demo-memory-map

Comment (AGlass0fMilk) [reply mbed_app.json:50]: Oh, perhaps the 4kB region is for the TLV trailers?

Comment (LDong-Arm) [reply mbed_app.json:50]: Yes I left some space for TLV (4KB should be way more than enough), but didn't precisely calculate its exact size which seems to depend on the configuration (e.g. if swap is enabled, etc.).

**Pull request merged**